oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/function/ServerToNodeMetadata.java

```diff
@@ -0,0 +1,155 @@
+import org.apache.jclouds.oneandone.rest.domain.Hdd;
+import org.apache.jclouds.oneandone.rest.domain.Server;
+import org.apache.jclouds.oneandone.rest.domain.ServerIp;
+import org.apache.jclouds.oneandone.rest.domain.Types.ServerState;
+import org.jclouds.collect.Memoized;
+import org.jclouds.compute.domain.Hardware;
+import org.jclouds.compute.domain.HardwareBuilder;
+import org.jclouds.compute.domain.NodeMetadata;
+import org.jclouds.compute.domain.NodeMetadataBuilder;
+import org.jclouds.compute.domain.OperatingSystem;
+import org.jclouds.compute.domain.Processor;
+import org.jclouds.compute.domain.Volume;
+import org.jclouds.compute.functions.GroupNamingConvention;
+import org.jclouds.domain.Location;
+import static org.jclouds.location.predicates.LocationPredicates.idEquals;
+import org.jclouds.util.InetAddresses2;
+
+public class ServerToNodeMetadata implements Function<Server, NodeMetadata> {
+
+   private final Supplier<Set<? extends Location>> locations;
+   private final Function<Hdd, Volume> fnVolume;
+   private final Function<HardwareFlavour, Hardware> fnHardware;
+   private final OneAndOneApi api;
+
+   private final GroupNamingConvention groupNamingConvention;
+
+   @Inject
+   public ServerToNodeMetadata(Function<Hdd, Volume> fnVolume,
+           Function<HardwareFlavour, Hardware> fnHardware,
+           @Memoized Supplier<Set<? extends Location>> locations,
+           OneAndOneApi api,
+           GroupNamingConvention.Factory groupNamingConvention) {
+      this.locations = locations;
+      this.api = api;
+      this.fnHardware = fnHardware;
+      this.fnVolume = fnVolume;
+      this.groupNamingConvention = groupNamingConvention.createWithoutPrefix();
+   }
+
+   @Override
+   public NodeMetadata apply(final Server server) {
+      checkNotNull(server, "Null server");
+
+      DataCenter dataCenter = api.dataCenterApi().get(server.datacenter().id());
+      Location location = find(locations.get(), idEquals(dataCenter.countryCode().toLowerCase()));
+      Hardware hardware = null;
+      //check if the server was built on a hardware flavour(Fixed instance)
+      if (server.hardware().fixedInstanceSizeId() != null && !"0".equals(server.hardware().fixedInstanceSizeId())) {
+         HardwareFlavour flavour = api.serverApi().getHardwareFlavour(server.hardware().fixedInstanceSizeId());
+         hardware = fnHardware.apply(flavour);
+
+      } else {
+         //customer hardware
+         double size = 0d;
+         List<Volume> volumes = Lists.newArrayList();
+         List<Hdd> storages = server.hardware().hdds();
+         if (storages != null) {
+            for (Hdd storage : storages) {
+               size += storage.size();
+               volumes.add(fnVolume.apply(storage));
+            }
+         }
+         // Build hardware
+         String id = String.format("cpu=%f,ram=%d,disk=%f", server.hardware().coresPerProcessor(), (int) server.hardware().ram(), size);
```

Comment:
Generate the ID with [this](https://github.com/jclouds/jclouds/blob/master/compute/src/main/java/org/jclouds/compute/util/AutomaticHardwareIdSpec.java#L60).